File: 1036-rotting-oranges/rotting-oranges.java
class Solution {
    public int orangesRotting(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        int minutes = -1;
        int freshCount = 0;
        Queue<int[]> q = new ArrayDeque<>();
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 2) {
                    q.add(new int[]{i, j});
                } else if (grid[i][j] == 1) {
                    freshCount++;
                }
            }
        }

        if (freshCount == 0) return 0;

        while (!q.isEmpty()) {
            int size = q.size();
            minutes++;
            for (int k = 0; k < size; k++) {
                int[] curr = q.poll();
                if (grid[curr[0]][curr[1]] == 3) continue;
                if (grid[curr[0]][curr[1]] == 1) freshCount--;
                if (freshCount == 0) return minutes;
                grid[curr[0]][curr[1]] = 3;
                List<int[]> neighbors = getNeighbors(curr, grid);
                for (int[] neighbor : neighbors) {
                    if (grid[neighbor[0]][neighbor[1]] == 1) {
                        q.add(neighbor);
                    }
                }
            }
        }
        return freshCount == 0 ? minutes : -1;
    }

    private List<int[]> getNeighbors(int[] curr, int[][] grid) {
        List<int[]> result = new ArrayList<>();
        int[][] directions = new int[][] {{0, 1}, {0, -1}, {-1, 0}, {1, 0}};
        for (int[] dir : directions) {
            int nexti = curr[0] + dir[0];
            int nextj = curr[1] + dir[1];
            if (nexti < grid.length && nexti >= 0 && nextj < grid[0].length && nextj >= 0) {
                result.add(new int[] {nexti, nextj});
            }
        }
        return result;
    }
}

File: 1036-rotting-oranges/rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        m = len(grid)
        n = len(grid[0])
        dirs = ((0,1), (1,0), (0,-1), (-1,0))
        days = 0
        fresh = 0
        q = deque()
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 2:
                    q.append((i, j))
                elif grid[i][j] == 1:
                    fresh += 1
        if not fresh: return 0
        while q:
            size = len(q)
            for i in range(size):
                curr = q.popleft()
                for di in dirs:
                    new_pos = (curr[0] + di[0], curr[1] + di[1])
                    if 0 <= new_pos[0] < m and 0 <= new_pos[1] < n and grid[new_pos[0]][new_pos[1]] == 1:
                        q.append(new_pos)
                        grid[new_pos[0]][new_pos[1]] = 2
                        fresh -= 1
                        if not fresh: return days + 1
            days += 1
        return -1